src/encode.cpp

```diff
@@ -435,6 +435,64 @@ optional<string> encodeOp(State &st, mlir::tensor::ExtractSliceOp op) {
   return {};
 }
 
+template<>
+optional<string> encodeOp(State &st, mlir::tensor::InsertSliceOp op) {
+  vector<Expr> offsets, sizes, strides;
+  auto src1 = st.regs.get<Tensor>(op.getOperand(0));
+  auto src2 = st.regs.get<Tensor>(op.getOperand(1));
+  auto res = op.getResult();
+  auto srcType1 = op.getOperand(0).getType().dyn_cast<mlir::ShapedType>();
+  auto srcType2 = op.getOperand(1).getType().dyn_cast<mlir::ShapedType>();
+  auto resType = res.getType().dyn_cast<mlir::ShapedType>();
+  
+#define GET_OP(vec, ee) { \
+    for (auto s: op.getMixed ## ee()) { \
+      vec.push_back(s.is<mlir::Value>() ? \
+      st.regs.get<Index>(s.get<mlir::Value>()) : \
+      Index(s.get<mlir::Attribute>().dyn_cast<mlir::IntegerAttr>().getInt())); \
+    } \
+  }
+  GET_OP(strides, Strides);
+  GET_OP(sizes, Sizes);
+  GET_OP(offsets, Offsets);
+#undef GET_OP
+
+  assert(offsets.size() == sizes.size() && sizes.size() == strides.size() &&
+          strides.size() == (size_t)srcType1.getRank() &&
+          srcType1.getRank() == srcType2.getRank() &&
+          srcType2 == resType
+        );
+
+  vector<Expr> inIdxs, src1Idxs, dims;
+  inIdxs = Index::boundIndexVars(resType.getRank());
+
+  for (unsigned i = 0; i < srcType1.getRank(); i++) {
+    uint64_t v;
+    sizes[i].isUInt(v);
+    assert(srcType1.getDimSize(i) >= v || srcType1.getDimSize(i) == -1);
+  }
+  
+  for (unsigned i = 0; i < srcType2.getRank(); i++) {
+    dims.push_back(Index(srcType2.getDimSize(i)));
+    src1Idxs.push_back((inIdxs[i] - offsets[i]) / strides[i]);
+  }
+
+  Expr src2Mapping = src2.get(inIdxs).first;
+
+  Expr output = Expr::mkIte(((inIdxs[0] - offsets[0]) % strides[0]).isZero() &
+                          ((inIdxs[0] - offsets[0]) / strides[0]).ult(sizes[0]),
+                          src1.get(src1Idxs).first, src2Mapping);
+
+  for (unsigned i = 1; i < resType.getRank(); i++) {
+    output = Expr::mkIte(((inIdxs[i] - offsets[i]) % strides[i]).isZero() &
+                          ((inIdxs[i] - offsets[i]) / strides[i]).ult(sizes[i]),
+                          output, src2Mapping);
+  }
```

Comment:
What about representing the output as

`ite (the big conjunction, src1.get(src1Idxs).first, src2Mapping)`

rather than

`ite(ite(cond, ite(cond, src1.get(src1Idxs).first, src2Mapping), src2Mapping), src2Mapping)`?